src/components/body/ammo-body.js

```diff
@@ -0,0 +1,480 @@
+/* global Ammo,THREE */
+const AmmoDebugDrawer = require("ammo-debug-drawer");
+const threeToAmmo = require("three-to-ammo");
+
+//Activation States
+const ACTIVE_TAG = 1;
+const ISLAND_SLEEPING = 2;
+const WANTS_DEACTIVATION = 3;
+const DISABLE_DEACTIVATION = 4;
+const DISABLE_SIMULATION = 5;
+
+//CollisionFlags
+const CF_STATIC_OBJECT = 1;
+const CF_KINEMATIC_OBJECT = 2;
+const CF_NO_CONTACT_RESPONSE = 4;
+const CF_CUSTOM_MATERIAL_CALLBACK = 8; //this allows per-triangle material (friction/restitution)
+const CF_CHARACTER_OBJECT = 16;
+const CF_DISABLE_VISUALIZE_OBJECT = 32; //disable debug drawing
+const CF_DISABLE_SPU_COLLISION_PROCESSING = 64; //disable parallel/SPU processing
+
+function almostEqualsVector3(epsilon, u, v) {
+  return Math.abs(u.x - v.x) < epsilon && Math.abs(u.y - v.y) < epsilon && Math.abs(u.z - v.z) < epsilon;
+}
+
+function almostEqualsQuaternion(epsilon, u, v) {
+  return (
+    (Math.abs(u.x - v.x) < epsilon &&
+      Math.abs(u.y - v.y) < epsilon &&
+      Math.abs(u.z - v.z) < epsilon &&
+      Math.abs(u.w - v.w) < epsilon) ||
+    (Math.abs(u.x + v.x) < epsilon &&
+      Math.abs(u.y + v.y) < epsilon &&
+      Math.abs(u.z + v.z) < epsilon &&
+      Math.abs(u.w + v.w) < epsilon)
+  );
+}
+
+let AmmoBody = {
+  schema: {
+    loadedEvent: { default: "" },
+    mass: { default: 1 },
+    gravity: { type: "vec3", default: { x: 0, y: -9.8, z: 0 } },
+    linearDamping: { default: 0.01 },
+    angularDamping: { default: 0.01 },
+    linearSleepingThreshold: { default: 1.6 },
+    angularSleepingThreshold: { default: 2.5 },
+    angularFactor: { type: "vec3", default: { x: 1, y: 1, z: 1 } },
+    activationState: {
+      default: ACTIVE_TAG,
+      oneOf: [ACTIVE_TAG, ISLAND_SLEEPING, WANTS_DEACTIVATION, DISABLE_DEACTIVATION, DISABLE_SIMULATION]
+    },
```

Comment:
Oh this is cool 😍 what do they do? I think we should use strings, maybe camelcase versions of these constants' names, instead of numbers here.

Reply:
Using numbers because that's how it is defined in bullet, so they can be passed directly into `body.setActivationState`
These flags are used for object deactivation (a.k.a. sleeping/waking). ACTIVE_TAG is the default, it will progress to WANTS_DEACTIVATION then ISLAND_SLEEPING when object is decided to have to come to rest. DISABLE_DEACTIVATION prevents an object from going to sleep at all. I assume DISABLE_SIMULATION effectively removes the object from physics simulation entirely, but I haven't yet tested it.
You can see these different activation states by turning on debug rendering. The wireframe colors map as follows:
ACTIVE_TAG: White
ISLAND_SLEEPING: Green
WANTS_DEACTIVATION: Blue
DISABLE_DEACTIVATION: Red
DISABLE_SIMULATION: Yellow

see: https://user-images.githubusercontent.com/1821537/53670274-fcdcc600-3c2e-11e9-86cd-90c630e22dee.gif

Reply:
Thanks!

I don't think that being able to pass them directly into bullet is enough to justify asking users to put something illegible like `body="activationState: 3"` in their HTML. It's different for an imperative API where you can import the constants, which just happen to have integer values. I'd prefer to use a map to convert human-readable names to the bullet integer values.

Reply:
So I did wind up moving all these constants into constants.js so that they could be imported easily, however that indeed doesn't help in regards to declarative syntax. I am a bit torn because I don't like obscuring what the underlying data types are (same reason I prefer using quaternions over vector3s for rotations). I will think about it more but I will probably convert these constants to strings as you suggest.